### ISSUE TITLE
stdio fails when output redirected to a file

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2259]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)| @jyknight reported|
|-|-|
|Trac ID|trac#2259|
|Type|defect|
|Created|2006-12-04 23:44:18Z|
|Branch|https://github.com/twisted/twisted/tree/stdio-file-redirect-2259|

At least on linux. Behavior of other OSes may or may not vary. See attached test case.

It works as expected when running normally. However, when redirecting output to a file, connectionLost gets called immediately.

This is because os.read() throws OSError so enableReadHack=True, but then select always returns readable&writeable for files.

Fixing this properly requires not using that read hack, which means tracking in-use sockets separately from the readable/writable lists (aka [#7599](https://github.com/twisted/twisted/issues/7599) and [#11168](https://github.com/twisted/twisted/issues/11168)), so we can get close notification.

Attachments:

* [testcase.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/507/507ccf1d683cee3ba79bbc61648f431badea4635/c6fa543491d4da1185be2699b43835c0d5ed8473.py) (1002 bytes) - added by jknight on 2006-12-04 23:44:38Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__2259 2259
type__defect defect
reporter__jknight jknight
priority__normal normal
milestone__ 
branch__branches_stdio_file_redirect_2259 branches/stdio-file-redirect-2259
branch_author__jknight__exarkun jknight, exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1165275858000000 1165275858000000
changetime__1273854874000000 1273854874000000
version__None None
owner__ 
cc__spiv
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @jyknight|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_3"></a>|@jyknight commented|
|-|-|

Actually this is fixable by itself, possibly, by putting a:
```
if stat.S_ISFIFO(os.fstat(self.fileno()).st_mode):
```
around the bit which does enableReadHack.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve commented|
|-|-|

[#2636](https://github.com/twisted/twisted/issues/2636) was closed as a duplicate of this.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [28892]) Add a test for StandardIO when stdout is redirected to a file; apply foom's suggested (probably) fix

refs [#2259](https://github.com/twisted/twisted/issues/2259)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [28891]) Branching to 'stdio-file-redirect-2259'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_8"></a>|@itamarst set owner to @exarkun|
|-|-|

This review doesn't quite cover the merits of the fix, since I don't understand it, but see below:

1. Missing news fragment.
2. The test doesn't actually write to the file. It would be good if that test or an additional one actually wrote something and the contents were verified.
3. Reading through the existing code confuses me. In particular, after a ProcessWriter.writeSomeData(), if all the data was successfully written and the read hack is enabled you will start reading, and doRead will then cause CONNECTION_LOST *unconditionally*(??!). That doesn't make sense to me at all, and sounds like the cause of the bug, but this patch doesn't seem to address the issue. It seems like, from comments elsewhere in the code, that the "read hack" is attempting to read to see if a connection was lost, or something, but that's never done, it's only done if we have broken version of Linux, but the read hack (read() to check closeness) should probably be happening in all cases since its goal is to work around limitations of select() reactor. Or something.
4. I find this code very mystifying in general, so even if my interpretation is wrong I would suggest better docstrings may be in order.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [28893]) Write some bytes to stdout to make sure it's really still open

refs [#2259](https://github.com/twisted/twisted/issues/2259)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [28895]) Try to make the fifo checking code clearer

refs [#2259](https://github.com/twisted/twisted/issues/2259)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [28896]) Document enableReadHack and add a comment where it's used in writeSomoeData

refs [#2259](https://github.com/twisted/twisted/issues/2259)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun removed owner|
|-|-|

Thanks for the review!

> 1. Missing news fragment.

Added.

> 2. The test doesn't actually write to the file.

Added.

> 3. Reading through the existing code confuses me.

Not too surprising. :)

> In particular, after a ProcessWriter.writeSomeData(), if all the data was successfully written and the read hack is enabled you will start reading

So, for one thing, the read hack won't be enabled in the case being fixed here.  That's what the new `S_ISFIFO` check the branch adds is for: to make sure to disable the read hack.  So, this case won't trigger `startReading`.

> and doRead will then cause CONNECTION_LOST *unconditionally*(??!).

It would, if the read hack were enabled.  But since it's disabled, instead it just calls `stopReading`.  It is still possible to get into `doRead` even with this change, because `_posixstdio.StandardIO.__init__` calls `startReading` on the `ProcessWriter` it creates.  I'm not _exactly_ sure why it does this, and it might be redundant.  I want to leave it alone for this fix, though, since it doesn't appear to hurt.

> That doesn't make sense to me at all, and sounds like the cause of the bug, but this patch doesn't seem to address the issue.

That's sort of right.  But the patch addresses the issue by changing `enableReadHack` from `True` to `False` for the affected case, rather than trying to change the meaning of `enableReadHack`.

> but the read hack (read() to check closeness) should probably be happening in all cases since its goal is to work around limitations of select() reactor

All of the reactors presently have the limitation, because they (still) don't track/report readability separate from close.  So the read hack is necessary on all of them for certain file descriptors (not file descriptors for normal files though).  But the `select` call is only necessary on broken versions of linux where readability on a pipe _might not_ mean the pipe is closed.  Note that `doRead` returns `None` if `enableReadHack` is `True`, `brokenLinuxPipeBehavior` is `True`, but the fd is not reported as readable _and_ writeable by `select`.

> 4. I find this code very mystifying in general, so even if my interpretation is wrong I would suggest better docstrings may be in order. 

I added some more comments and docstrings and re-arranged the way `enableReadHack` is initialized.  Perhaps that makes things clearer.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

(In [28898]) Give MockOS an fstat function

refs [#2259](https://github.com/twisted/twisted/issues/2259)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_14"></a>|@itamarst commented|
|-|-|

What was confusing me was that I'd blanked on the fact that if doRead is called that means the fd is readable, so it's not an unconditional connection lost, it's only lost when fd is readable.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun set owner to @exarkun|
|-|-|

There are some issues with epoll and on Windows.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun commented|
|-|-|

Well, this approach totally fails for epollreactor.  epoll cannot be asked to monitor normal files.  It won't even lie like select() and poll(), it just fails.

I think that means this needs to be fixed by not using `ProcessWriter` at all in `StandardIO` if the stdout fd is a regular file.  That would mean `ProcessWriter` doesn't actually need to change at all.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun commented|
|-|-|

Considering this further, the epoll behavior in this branch isn't a regression.  trunk is as broken for epoll+normal-file-stdout.  So that's not necessarily a reason to block applying this solution (though of course the new tests would need to be marked todo for epoll reactor, and there should be a new ticket for making this work with epoll).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun commented|
|-|-|

(In [28907]) todo the new test on epollreactor

refs [#2259](https://github.com/twisted/twisted/issues/2259)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_19"></a>|@exarkun removed owner|
|-|-|

Alright, Windows and epollreactor now skip the new test.  If this seems like a reasonable course, I'll file a new ticket for making this work on epoll and reference it from the skip.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_20"></a>|@itamarst commented|
|-|-|

If using a thread for stdout would make it work everywhere (and also solve issues when talking to programs that don't expect their stdin to be non-blocking), maybe we should just do that and fix everything in one fell swoop?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_21"></a>|@glyph commented|
|-|-|

It's never as easy as "spawn a thread and write 'till you block" :).  If we use a thread for stdout, we need to make sure that it's still possible to cleanly shut down without crashing python or getting into any race conditions or whatever, even if the FD being used for stdout starts off blocking and never recovers.  It also adds some overhead if you've got a server with a bunch of subprocesses that it needs to talk to over stdin/stdout.

That _could_ still be the right thing to do, of course, but it would be nice to avoid it in the cases where we could.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_22"></a>|@itamarst commented|
|-|-|

Threads are easy like Monday morning, but OK. What's the epoll-specific solution, if any?

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_23"></a>|@jyknight commented|
|-|-|

Don't forget about [#3442](https://github.com/twisted/twisted/issues/3442)...

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_24"></a>|@exarkun commented|
|-|-|

If someone wants to implement the thread-based solution, I'm not going to stand in their way.  I'm not likely to implement it in the near future.  I'm interested in having this fixed in the selectreactor case because it's breaking a real application.  That application handles three file descriptors, so it doesn't really need epoll.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_25"></a>|@glyph commented|
|-|-|

Replying to [jknight](#note_23):
> Don't forget about [#3442](https://github.com/twisted/twisted/issues/3442)...

Well, I'm _not_ forgetting about it, but I sure wish I could.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_26"></a>|@itamarst set owner to @exarkun|
|-|-|

1. You should open ticket for epoll and then update the #NNNN reference in the skip message for the test. Also maybe update news fragment to mention epoll is still issue.

2. Is there any way we can detect epoll when creating the StandardIO class, and throwing exception to the user so they don't get really weird result and have to track the bug down? If you think fixing epoll soon is realistic that may be unnecessary.

Or maybe we can go straight to fixing 3442 before next release ;)

Other than that, I guess it's done.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_27"></a>|@exarkun removed owner|
|-|-|

> 1. You should open ticket for epoll and then update the #NNNN reference in the skip message for the test. Also maybe update news fragment to mention epoll is still issue.

Done.

> 2. Is there any way we can detect epoll when creating the StandardIO class, and throwing exception to the user so they don't get really weird result and have to track the bug down?

Yea-ish.  It raises a `RuntimeError` with a message and a ticket reference now.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_28"></a>|@itamarst set owner to @exarkun|
|-|-|

EPERM could in theory be caused by some other reactor with some other arbitrary limitation; I suggest slightly expanding the RuntimeError message, to say something like "The current reactor does not support this type of file descriptor, e.g. epollreactor doesn't support etc.".

Other than that, looks good, please commit.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_29"></a>|@exarkun set status to `closed`|
|-|-|

(In [28927]) Merge stdio-file-redirect-2259

Author: jknight, exarkun
Reviewer: itamar
Fixes: [#2259](https://github.com/twisted/twisted/issues/2259)

Support stdout being redirected to a normal file when using `twisted.internet.stdio`.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
